Merging Python Tutor + Shapes library (#1)
joshpoll/bluefish-docs: classic/docs/example-gallery/python-tutor.md

# Python Tutor: A Bluefish Case Study

## Introduction

In this section, we will explore how to use Bluefish to build complex domain specific diagrams like Python Tutor visualizations.

First, let's breakdown the parts of a Python Tutor visualization. Below is an example of a Python Tutor diagram that we will recreate in this section:

![Example Python Tutor Diagram](../images/python-tutor.png)

We can divide the components of this diagram into two groups: the components associated with the Global Frame and the Objects. Let's build these two parts of the diagram separately, and then compose the two parts to create the entire Python Tutor visualization.

## The Global Frame

> The Global Frame itself is a composition of various other components. In particular, the Global Frame contains some text components, some variables, and some components responsible for the styling if the output.

Reviewer: ```suggestion
The Global Frame itself is a composition of various other components. In particular, the Global Frame contains some text components, some variables, and some components responsible for the styling of the output.
```

          fontSize={"16px"}
          fill={"grey"}
        />

        <Align center>
          <Ref to={valueRef} />
          <Ref to={itemRef} />
        </Align>

        <Align left to={"centerRight"}>
          <Ref to={boxRef} />
          <Ref to={itemRef} />
        </Align>

        <Align topLeft>
          <Ref to={oneRef} />
          <Ref to={boxRef} />
        </Align>
      </Group>

      <Space vertically by={10}>
        <Ref to={labelRef} />
        <Ref to={elemRef} />
      </Space>
    </Group>
  );
});

render(
  <SVG width={500} height={100}>
    <Objects
      nextObject={{ opId: "object2" }}
      objectType={"tuple"}
      value={"1"}
      opId={"object1"}
    />
  </SVG>
);
```

We can imagine that the objects are located in a grid, and we can specify the contents of each grid with a series of `Row` components. Therefore, our set of Objects is essentially a matrix.

```tsx live noInline
const ObjectsMatrix = forwardRef(function _ObjectsMatrix(
  { objects, rows },
  ref
) {
  // lookup map for the yellow objects
  const objMap = new Map();
  objects.forEach((obj) => objMap.set(obj.opId, obj));

  return (
    <Group ref={ref} name={"matrix-rows"}>

>       <Space name={"rowSpace"} vertically by={50}>

Reviewer: Should this be a Col? I'm not sure if the Row positions are actually specified. (I assume they're currently left-aligning automatically since their positions are not set, but we should try not to rely on this behavior.)

      </Space>

      {objectLinks.map((link) => (
        <Group>
          <Link {...link} />
        </Group>
      ))}

      {variableLinks.map((link) => (
        <Group>
          <Link {...link} />
        </Group>
      ))}
    </Group>
  );
});

render(
  // Try changing the objects or the arrangement of the matrix!
  <SVG width={800} height={400}>
    <PythonTutor
      variables={[
        { pointObject: { opId: "o1" }, value: "", name: "c", opId: "v1" },
        { pointObject: { opId: "o2" }, value: "", name: "d", opId: "v2" },
        { pointObject: null, name: "x", value: "5", opId: "v3" },
      ]}
      opId={"pythonTutorFrame"}
      objects={[
        {
          nextObject: { opId: "o2" },
          objectType: "tuple",
          value: "1",
          opId: "o1",
        },
        {
          nextObject: { opId: "o3" },
          objectType: "tuple",
          value: "2",
          opId: "o2",
        },
        { nextObject: null, objectType: "tuple", value: "3", opId: "o3" },
      ]}
      rows={[
        { depth: 0, nodes: ["", "o2", "o3"] },
        { depth: 1, nodes: ["o1", "", ""] },
      ]}
    />
  </SVG>
);
```

## Generalizing Python Tutor

> Python Tutor diagrams are very flexible, and the content represented in the diagram can be very diverse. For example, objects could have arbitrary length. Objects themselves could simply store pointers to other primitive types. In the following sections, we will explore how to generalize the Bluefish Python Tutor diagram to flexibility handle these situations.

Reviewer: ```suggestion
Python Tutor diagrams are very flexible, and the content represented in the diagram can be very diverse. For example, objects could have arbitrary length. Objects themselves could simply store pointers to other primitive types. In the following sections, we will explore how to generalize the Bluefish Python Tutor diagram to handle these situations.
```

  const zeroRef = useRef(null);
  const oneRef = useRef(null);

  const fontFamily = "verdana, arial, helvetica, sans-serif";

  return (
    <Group ref={ref} name={objectId}>
      <Text
        ref={objectTypeRef}
        contents={objectType}
        fontFamily={fontFamily}
        fontSize={"16px"}
        fill={"grey"}
      />

      {/* separate names for each rectangle so that the arrow can go from the center of pointer to the center left of pointed */}
      <Group ref={objectRef}>
        <Rect
          ref={boxOneRef}
          name={`elm_1_${objectId}`}
          height={60}
          width={70}
          fill={"#ffffc6"}
          stroke={"grey"}
        />
        <Rect
          ref={boxZeroRef}
          name={`elm_0_${objectId}`}
          height={60}
          width={70}
          fill={"#ffffc6"}
          stroke={"grey"}
        />

        {/* Generate Text in Object based on whether it is a string or a pointer */}
        {objectValues[0].type == "string" ? (
          <Text
            ref={boxZeroValueRef}
            contents={objectValues[0].value}
            fontSize={"24px"}
            fill={"black"}
          />
        ) : (

>           <Text ref={boxZeroValueRef} contents={""} fill={"none"} />

Reviewer: Does something break if e.g. this were a 0-width, 0-height rect? I wonder if we need a `null`/`Fragment` Bluefish component so that Ref resolution still works even though we don't want to render anything...

Author: I'm not sure if this is what you mean - but if I change either the object associated with boxOneRef or boxZeroRef, the diagram will still render and there's no error. It just looks a bit weird because the background yellow box is missing, so the text does not have anything to align with.

        { pointObject: { opId: "o2" }, value: "", name: "d", opId: "v2" },
        { pointObject: null, name: "x", value: "5", opId: "v3" },
      ]}
      opId={"pythonTutorFrameV2"}
      objects={[
        {
          objectType: "tuple",
          objectId: "o1",
          objectValues: [
            { type: "string", value: "1" },
            { type: "pointer", pointId: "o2" },
          ],
        },
        {
          objectType: "tuple",
          objectId: "o2",
          objectValues: [
            { type: "pointer", pointId: "o4" },
            { type: "pointer", pointId: "o3" },
          ],
        },
        {
          objectType: "tuple",
          objectId: "o3",
          objectValues: [
            { type: "string", value: "3" },
            { type: "string", value: "6" },
          ],
        },
        {
          objectType: "tuple",
          objectId: "o4",
          objectValues: [
            { type: "string", value: "5" },
            { type: "string", value: "7" },
          ],
        },
      ]}
      rows={[
        { depth: 0, nodes: ["", "o2", "o3"] },
        { depth: 1, nodes: ["o1", "", "o4"] },
      ]}
    />
  </SVG>
);
```

> ### Creating Objects of Arbitrarily Length

Reviewer: ```suggestion
### Creating Objects of Arbitrary Length
```

      <Rect
        name={`elmBox_${tupleIndex}_${objectId}`}
        height={60}
        width={70}
        fill={"#ffffc6"}
        stroke={"grey"}
      />
      <Text
        name={`elmLabel_${tupleIndex}_${objectId}`}
        contents={`${tupleIndex}`}
        fontFamily={fontFamily}
        fontSize={"16px"}
        fill={"grey"}
      />
      {tupleData.type == "string" ? (
        <Text
          name={`elmVal_${tupleIndex}_${objectId}`}
          contents={tupleData.value}
          fontSize={"24px"}
          fill={"black"}
        />
      ) : (
        <Text
          name={`elmVal_${tupleIndex}_${objectId}`}
          contents={""}
          fill={"none"}
        />
      )}

      <Align center>
        <Ref to={`elmVal_${tupleIndex}_${objectId}`} />
        <Ref to={`elmBox_${tupleIndex}_${objectId}`} />
      </Align>

      <Align topLeft>
        <Ref to={`elmLabel_${tupleIndex}_${objectId}`} />
        <Ref to={`elmBox_${tupleIndex}_${objectId}`} />
      </Align>
    </Group>
  );
});

render(
  <SVG width={100} height={100}>
    <ElmTuple
      tupleIndex={"0"}
      tupleData={{ type: "string", value: "3" }}
      objectId={"object1"}
    />
  </SVG>
);
```

> Now that we've refactored the tuple element into a Bluefish element, let's rewrite Objects so that it can be arbitrary lengths.

Reviewer: ```suggestion
Now that we've refactored the tuple element into a Bluefish element, let's rewrite Objects so that it can have an arbitrary length.
```

```tsx live noInline
const ObjectsV3 = forwardRef(function _Objects({ objectType, objectValues, objectId }, ref) {
  // objectValues: list of values for different parts of object

  const objectTypeRef = useRef(null);
  const objectRef = useRef(null);

  const elementWithIndex = objectValues.map((elementVals, index) => {
    return { ...elementVals, order: index };
  });

>   const allExceptElmZero = elementWithIndex.filter((elementData) => {

Reviewer: It doesn't seem like this or `elementWithIndex` are used anymore?

Author: Oops yeah forgot to delete this one after refactoring the object / element tuple. Should be removed now.
